### ISSUE TITLE
cmake error on CentOS 7.1

### ISSUE THREAD

**Issue by assassin5615** [google/benchmark#773]
This is probably similar to #351 . 
just followed the instructions in README and run into following error while running cmake

```
Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY - Failed
-- Performing Test HAVE_CXX_FLAG_COVERAGE
-- Performing Test HAVE_CXX_FLAG_COVERAGE - Success
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX -- compiled but failed to run
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX -- compiled but failed to run
CMake Error at CMakeLists.txt:250 (message):
  Failed to determine the source files for the regular expression backend
```
The solution is the same as in #351 ,  -DCMAKE_CROSSCOMPILING=1  -DRUN_HAVE_STD_REGEX=0 -DRUN_HAVE_POSIX_REGEX=0 needs to be added to fix the issue.

but what I don't get is that how this is related to cross compiling.  is this error always expected on non Debian and ubuntu OS? thanks

**Comment by LebedevRI** [comment]
How do they fail? (see one of the `CMake*.txt` files it mentions at the end)
Maybe some `-dev` package is missing?

**Comment by assassin5615** [comment]
I assume that you mean CMake*.log? I didn't find anything useful. contents at the end of CMakeOutput.log is as below, it's only for HAVE_CXX_FLAG_COVERAGE which is before the REGEX check.

```
Source file was:
int main() { return 0; }
Performing C++ SOURCE FILE Test HAVE_CXX_FLAG_COVERAGE succeded with the following output:
Change Dir: /home/ylei/development/third_party/benchmark/build/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/gmake" "cmTC_a413e/fast"
/usr/bin/gmake -f CMakeFiles/cmTC_a413e.dir/build.make CMakeFiles/cmTC_a413e.dir/build
gmake[1]: Entering directory `/home/ylei/development/third_party/benchmark/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_a413e.dir/src.cxx.o
/home/ylei/usr/local/gcc-7.3.0/bin/g++     -std=c++11  -Wall  -Wextra  -Wshadow  -pedantic  -pedantic-errors  -fstrict-aliasing  -Wno-deprecated-declarations  -Wstrict-aliasing -DHAVE_CXX_FLAG_COVERAGE  --coverage   --coverage -o CMakeFiles/cmTC_a413e.dir/src.cxx.o -c /home/ylei/development/third_party/benchmark/build/CMakeFiles/CMakeTmp/src.cxx
Linking CXX executable cmTC_a413e
/usr/local/bin/cmake -E cmake_link_script CMakeFiles/cmTC_a413e.dir/link.txt --verbose=1
/home/ylei/usr/local/gcc-7.3.0/bin/g++     -std=c++11  -Wall  -Wextra  -Wshadow  -pedantic  -pedantic-errors  -fstrict-aliasing  -Wno-deprecated-declarations  -Wstrict-aliasing -DHAVE_CXX_FLAG_COVERAGE  --coverage    CMakeFiles/cmTC_a413e.dir/src.cxx.o  -o cmTC_a413e -rdynamic
gmake[1]: Leaving directory `/home/ylei/development/third_party/benchmark/build/CMakeFiles/CMakeTmp'

Source file was:
int main() { return 0; }
```


Contents at the end of CMakeError.log as below. it's for THREAD_SAFETY which is also before REGEX check.


```
Source file was:
int main() { return 0; }
Performing C++ SOURCE FILE Test HAVE_CXX_FLAG_WTHREAD_SAFETY failed with the following output:
Change Dir: /home/ylei/development/third_party/benchmark/build/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/gmake" "cmTC_a5ac9/fast"
/usr/bin/gmake -f CMakeFiles/cmTC_a5ac9.dir/build.make CMakeFiles/cmTC_a5ac9.dir/build
gmake[1]: Entering directory `/home/ylei/development/third_party/benchmark/build/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_a5ac9.dir/src.cxx.o
/home/ylei/usr/local/gcc-7.3.0/bin/g++     -std=c++11  -Wall  -Wextra  -Wshadow  -pedantic  -pedantic-errors  -fstrict-aliasing  -Wno-deprecated-declarations  -Wstrict-aliasing -DHAVE_CXX_FLAG_WTHREAD_SAFETY  -Wthread-safety   -Wthread-safety -o CMakeFiles/cmTC_a5ac9.dir/src.cxx.o -c /home/ylei/development/third_party/benchmark/build/CMakeFiles/CMakeTmp/src.cxx
g++: error: unrecognized command line option '-Wthread-safety'; did you mean '-fthread-jumps'?
g++: error: unrecognized command line option '-Wthread-safety'; did you mean '-fthread-jumps'?
gmake[1]: *** [CMakeFiles/cmTC_a5ac9.dir/src.cxx.o] Error 1
gmake[1]: Leaving directory `/home/ylei/development/third_party/benchmark/build/CMakeFiles/CMakeTmp'
gmake: *** [cmTC_a5ac9/fast] Error 2

Source file was:
int main() { return 0; }
```
it does not look like that -dev package is missing. as with -DCMAKE_CROSSCOMPILING=1 -DRUN_HAVE_STD_REGEX=0 -DRUN_HAVE_POSIX_REGEX=0 added, I got following output that says REGEX is successful.

```
-- Performing Test HAVE_STD_REGEX -- success
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX -- success
-- Performing Test HAVE_STEADY_CLOCK
-- Performing Test HAVE_STEADY_CLOCK
-- Performing Test HAVE_STEADY_CLOCK -- success
```

**Comment by Jerry-Ma** [comment]
I am suspecting that I am encountering the same problem.

```
benchmark_SOURCE_DIR: /home/zhiyuan.ma/toltec/bench_pca_cpu_gpu/build/_deps/benchmark-src
benchmark_BINARY_DIR: /home/zhiyuan.ma/toltec/bench_pca_cpu_gpu/build/_deps/benchmark-build
-- Failed to find LLVM FileCheck
-- Found Git: /usr/bin/git (found version "1.8.3.1")
-- git Version: v1.5.0-309de598-dirty
-- Version: 1.5.0
-- Performing Test HAVE_CXX_FLAG_STD_CXX11
-- Performing Test HAVE_CXX_FLAG_STD_CXX11 - Success
-- Performing Test HAVE_CXX_FLAG_WALL
-- Performing Test HAVE_CXX_FLAG_WALL - Success
-- Performing Test HAVE_CXX_FLAG_WEXTRA
-- Performing Test HAVE_CXX_FLAG_WEXTRA - Success
-- Performing Test HAVE_CXX_FLAG_WSHADOW
-- Performing Test HAVE_CXX_FLAG_WSHADOW - Success
-- Performing Test HAVE_CXX_FLAG_WERROR
-- Performing Test HAVE_CXX_FLAG_WERROR - Success
-- Performing Test HAVE_CXX_FLAG_PEDANTIC
-- Performing Test HAVE_CXX_FLAG_PEDANTIC - Success
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS
-- Performing Test HAVE_CXX_FLAG_PEDANTIC_ERRORS - Success
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32
-- Performing Test HAVE_CXX_FLAG_WSHORTEN_64_TO_32 - Failed
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_FSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED_DECLARATIONS - Success
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED
-- Performing Test HAVE_CXX_FLAG_WNO_DEPRECATED - Success
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING
-- Performing Test HAVE_CXX_FLAG_WSTRICT_ALIASING - Success
-- Performing Test HAVE_CXX_FLAG_WD654
-- Performing Test HAVE_CXX_FLAG_WD654 - Failed
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY
-- Performing Test HAVE_CXX_FLAG_WTHREAD_SAFETY - Failed
-- Performing Test HAVE_CXX_FLAG_COVERAGE
-- Performing Test HAVE_CXX_FLAG_COVERAGE - Success
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX
-- Performing Test HAVE_STD_REGEX -- compiled but failed to run
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX
-- Performing Test HAVE_POSIX_REGEX -- compiled but failed to run
CMake Error at build/_deps/benchmark-src/CMakeLists.txt:252 (message):
  Failed to determine the source files for the regular expression backend
```

```
$ cmake --version
cmake version 3.15.4
$ uname -r
3.10.0-862.el7.x86_64
```

And I tried two different compilers, same problem:
```
$ icpc -V
Intel(R) C++ Intel(R) 64 Compiler for applications running on Intel(R) 64, Version 19.0.0.117 Build 20180804
Copyright (C) 1985-2018 Intel Corporation.  All rights reserved.
$ /home/opt_software/gcc-8.1.0/bin/gcc -v
Using built-in specs.
COLLECT_GCC=/home/opt_software/gcc-8.1.0/bin/gcc
COLLECT_LTO_WRAPPER=/home/opt_software/gcc-8.1.0/bin/../libexec/gcc/x86_64-pc-linux-gnu/8.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: ./configure --prefix=/opt/software/gcc-8.1.0/ --disable-multilib
Thread model: posix
gcc version 8.1.0 (GCC)
[zhiyuan.ma@sc-nm build]$ /home/opt_software/gcc-8.1.0/bin/g++ -v
Using built-in specs.
COLLECT_GCC=/home/opt_software/gcc-8.1.0/bin/g++
COLLECT_LTO_WRAPPER=/home/opt_software/gcc-8.1.0/bin/../libexec/gcc/x86_64-pc-linux-gnu/8.1.0/lto-wrapper
Target: x86_64-pc-linux-gnu
Configured with: ./configure --prefix=/opt/software/gcc-8.1.0/ --disable-multilib
Thread model: posix
gcc version 8.1.0 (GCC)
```

**Comment by Luthaf** [comment]
I had the same problem, and after digging around a bit (using `cmake --debug-trycompile`), this is an aftermath of the changes in #50 / #58 

The default gcc on CentOS 7 is gcc 4.8, which default to `-std=c++98`, meaning that the `nullptr` in `cmake/posix_regex.cpp` causes the compilation to fail:

https://github.com/google/benchmark/blob/c50ac68c50ff8da3827cd6720792117910d85666/cmake/posix_regex.cpp#L10

A hacky way to solve this is to force the use of `-std=c++11` in try_compile, for example with `set(CMAKE_REQUIRED_DEFINITIONS "-std=c++11")`. 

A less hacky way would be to revert changes from #58, and instead use `#pragma GCC diagnostic ignored "-Wzero-as-a-null-pointer-constant"`, or disable `-Werror` in try_compile altogether.

Ping @dominichamon from the PR that introduced these changes.

**Comment by assassin5615** [comment]
> CMAKE_REQUIRED_DEFINITIONS "-std=c++11").

I used gcc 7.3 and still encountered this problem. I also tried your set(CMAKE_REQUIRED_DEFINITIONS "-std=c++11") solution, but it doesn't work.

**Comment by Luthaf** [comment]
gcc 7.3 should have std::regex, so what you see might be a different problem.

**Comment by assassin5615** [comment]
> gcc 7.3 should have std::regex, so what you see might be a different problem.

@Luthaf yes, you are right. the error is actually different when using gcc7.3, it says compiled but failed to run. I did some investigation, it's because that my LD_LIBRARY_PATH is not set correctly.

once gcc7.3 works, I also tried gcc 4.8.5, it works too with export  CMAKE_CXX_FLAGS="-std=c++11" 
you are absolutely right about this issue. thanks a lot for clarifying this problem.
